### PR TITLE
added task-endpoint enhancement

### PR DESCRIPTION
**Pull request by vinayak-trivedi**
- enhances the currently available task assignment endpoint based on the new requirements

### CODE REVIEW

File: test/integration/tasks.test.js [hunk 1]
```diff
@@ -476,38 +476,6 @@ describe("Tasks", function () {
...
-
-      const res = await chai
-        .request(app)
-        .patch(`/tasks/self/${taskId}?dev=true`)
-        .set("cookie", `${cookieName}=${jwt}`)
-        .send(dataToSend);
-
-      expect(res).to.have.a.status(200);
-      expect(res.body.message).to.be.equal("task updated and another task got assigned");
-    });
```
Comment (vinayak-trivedi) [inline test/integration/tasks.test.js:510]: this is removed as the logic for task assignment has changed, and we are not going to do it it on `/task/self/:id` endpoint

File: test/integration/tasks.test.js [hunk 2]
```diff
@@ -536,30 +504,4 @@ describe("Tasks", function () {
...
-      const res = await chai
-        .request(app)
-        .patch(`/tasks/assign/self`)
-        .set("cookie", `${cookieName}=${superUserJwt}`)
-        .send();
-
-      expect(res).to.have.status(200);
-      expect(res.body.message).to.be.equal("Task assigned");
-    });
-  });
```
Comment (vinayak-trivedi) [inline test/integration/tasks.test.js:564]: will write the tests in another PR after @DashDeipayan's PR get merged

**Comment by alishvagadiya** [comment]
Please add more description, it will help reviewers, and it will help get more context about PR.

**Review by ivinayakg — APPROVED** [review]
The code looks good.
A couple of concerns :-
- Although there is code added there are no new tests, is there a specific reason?
- Added a comment check that out

**Review by heyrandhir — DISMISSED** [review]
LGTM

File: utils/users.js [hunk 3]
```diff
@@ -81,9 +81,24 @@ const getParticipantUserIds = async (participantArray) => {
...
 
+function getLowestLevelSkill(skills) {
+  let minLevel = skills[0].level;
+  let skill;
+  let level;
+  for (const skillfield of skills) {
+    if (skillfield.level < minLevel) {
+      minLevel = skillfield.level;
+      skill = skillfield.skill;
+      level = skillfield.level;
```
[the remark this answers is not in the document]
Comment (vinayak-trivedi) [reply utils/users.js:92]: - I will be writing test once @DashDeipayan PR gets merged

File: models/tasks.js [hunk 4]
```diff
@@ -194,13 +195,29 @@ const fetchUserTasks = async (username, statuses = [], field, order) => {
...
+  if (!availableTasks.empty) {
+    availableTasks.forEach((item) => idArray.push(item.id));
+
+    if (!skill) {
+      task = await ItemModel.where("itemtype", "==", "TASK")
+        .where("itemid", "in", idArray)
+        .where("levelnumber", ">=", 2)
+        .limit(1)
+        .get();
+    } else {
```
Comment (prakashchoudhary07) [inline models/tasks.js:211]: What is the max size of idArray
And when filtering based on  `in idArray`, I think `in` combines up to 10 equality, So if idArray size is more  than 10 then what will it happen?

Ref: https://firebase.google.com/docs/firestore/query-data/queries#in_not-in_and_array-contains-any

File: models/tasks.js [hunk 5]
```diff
@@ -194,13 +195,29 @@ const fetchUserTasks = async (username, statuses = [], field, order) => {
...
+      task = await ItemModel.where("tagname", "==", skill)
+        .where("itemtype", "==", "TASK")
+        .where("itemid", "in", idArray)
+        .where("levelnumber", ">=", level)
```
Comment (prakashchoudhary07) [inline models/tasks.js:215]: RE: 'in'

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

**Review by prakashchoudhary07 — DISMISSED** [review]

File: models/tasks.js [hunk 6]
```diff
@@ -193,29 +194,45 @@ const fetchUserTasks = async (username, statuses = [], field, order) => {
...
+      return {
+        task: {
+          id,
+          ...taskData,
+        },
+      };
+    }
+    return { taskNotFound: true };
+  }
   return { taskNotFound: true };
```
Comment (prakashchoudhary07) [inline models/tasks.js:236]: Task not found two time?

File: models/tasks.js [hunk 7]
```diff
@@ -193,29 +194,45 @@ const fetchUserTasks = async (username, statuses = [], field, order) => {
...
+
+    if (!skill) {
+      task = await ItemModel.where("itemType", "==", "TASK").where("levelNumber", "<=", 2).get();
+    } else {
+      task = await ItemModel.where("tagName", "==", skill)
+        .where("itemType", "==", "TASK")
+        .where("levelNumber", ">=", level)
+        .where("levelNumber", "<=", level + 2)
+        .get();
+    }
```
Comment (prakashchoudhary07) [inline models/tasks.js:214]: Are we not limiting the data here?

File: models/tasks.js [hunk 8]
```diff
@@ -226,9 +243,24 @@ const getNewTask = async (skill, level) => {
  * @returns {Promise<task>|object}
  */
 
-const fetchSkillLevelTask = async (skill, level) => {
+const fetchSkillLevelTask = async (userId) => {
   try {
-    const task = await getNewTask(skill, level);
+    let task;
+    const data = await ItemModel.where("itemId", "==", userId).where("tagType", "==", "SKILL").limit(10).get();
```
Comment (prakashchoudhary07) [inline models/tasks.js:249]: Won't this always return the first 10 data only, how will we get the next 10?

File: models/tasks.js (hunk 6 above)
Comment (vinayak-trivedi) [reply models/tasks.js:236]: yes if there is no available task then also we want to return that, and if there was available but we couldn't find anything we will return taskNotFound

File: models/tasks.js (hunk 7 above)
Comment (vinayak-trivedi) [reply models/tasks.js:214]: no, as discussed I will be doing enhancement in other PR

File: models/tasks.js (hunk 8 above)
Comment (vinayak-trivedi) [reply models/tasks.js:249]: mostly 10 will do the work for us, we do not want all the skills of the user

Comment (prakashchoudhary07) [reply models/tasks.js:249]: Won't this return the same skill each and every time?

File: models/tasks.js (hunk 7 above)
Comment (prakashchoudhary07) [reply models/tasks.js:214]: Can you put the issue link here please 😄

Comment (vinayak-trivedi) [reply models/tasks.js:214]: https://github.com/Real-Dev-Squad/website-backend/issues/843

**Review by prakashchoudhary07 — APPROVED** [review]

**Pull request merged**